### ISSUE TITLE
Refactor return type of `BlockChainClient::code`

### ISSUE THREAD

**Issue by 0x7CFE** [paritytech/parity-ethereum#7098]
Currently it is defined like this:

```Rust
/// Get address code at given block's state.
fn code(&self, address: &Address, id: BlockId) -> Option<Option<Bytes>>;
```

Technically it's OK, but may be the source of confusion or even misinterpretation during code review. Taking into account potential importance of this method, I may suggest refactoring return type to something more verbose.

Because absence of code is not a error per se, we should not use `Result` here. We may introduce an `enum` that would either return a value or provide a reason why it wasn't there. As far as I understand, value may be missing due to nonexistent address, or due to actual absence of the code. 

So, it may look something like this:
```Rust
enum AddressCode {
    AddressNotFound,
    Empty,
    Some(Bytes),
}
```

**Comment by tomusdrw** [comment]
In `Option<Option<Bytes>>` `None` is returned when the state is pruned, so it's more like:
```
enum StateResult<T> {
  StateMissing,
  Some(T),
}
```
and then `fn code() -> StateResult<Option<Bytes>>` since the pattern will be quite common in all other methods in client.

**Comment by folsen** [comment]
@0x7CFE Is this addressed in #7038? If so please close.

**Comment by 0x7CFE** [comment]
> Is this addressed in #7038?

@folsen No, it isn't.
